### PR TITLE
PR with code for the blog "Anomaly detection using embeddings and GenAI"

### PR DESCRIPTION
**Pull request by kwulffert23**
PR with code for the blog "Anomaly detection using embeddings and GenAI" as part of the Blogathon initiative H2 of this year.

### CODE REVIEW

File: genai-anomaly-detection/1_anomaly_detection_pca.py [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+# MAGIC %md
+# MAGIC # Anomaly detection on embeddings
+
+# COMMAND ----------
+
+# MAGIC %md
+# MAGIC In this notebook, we will explore Principal Component Analysis (PCA) as a method for identifying outliers in product descriptions within vendors by analyzing the embeddings of these descriptions. 
+# MAGIC <br>PCA offers a robust technique for dimensionality reduction, which can potentially reveal anomalies based on reconstruction errors in the embedding space. 
+# MAGIC <br>It is crucial to recognize that PCA is just one of many available algorithms for outlier detection, each with its own set of advantages and limitations. Alternatives like t-SNE, UMAP, or nearest neighbors may provide different insights depending on the data and the specific characteristics of the embeddings. Thus, while PCA can be effective, it is not a silver bullet, and a comprehensive evaluation of multiple techniques is essential for robust anomaly detection.
+
```
Comment (marzervou) [inline genai-anomaly-detection/1_anomaly_detection_pca.py:11]: Add a readme with cluster configurations and ML runtime needs?

File: genai-anomaly-detection/2_hybrid_function-calling-anomaly-examples.py [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+            results.append(r)
+        return results
+
+
+def results_to_dataframe(products: List[str], responses: List[str]):
+    """Combines reviews and model responses into a dataframe for tabular display."""
+    return pd.DataFrame({"Products list": products, "Model response": responses})
+
+
+def list_to_dataframe(elements):
```
Comment (marzervou) [inline genai-anomaly-detection/2_hybrid_function-calling-anomaly-examples.py:125]: Seems that this function is not being used anywhere, maybe drop?

File: genai-anomaly-detection/1_anomaly_detection_pca.py [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+# COMMAND ----------
+
+# MAGIC %md
+# MAGIC ## Data preparation
+
+# COMMAND ----------
+
+import pandas as pd
+df = pd.read_csv('./data/purchase_vendors_products_synthetic.csv')
```
Comment (marzervou) [inline genai-anomaly-detection/1_anomaly_detection_pca.py:20]: I understand you did this for simplicity, but i'd recommend just loading UC tables across the notebook

File: genai-anomaly-detection/1_anomaly_detection_pca.py (hunk 1 above)
Comment (kwulffert23) [reply genai-anomaly-detection/1_anomaly_detection_pca.py:11]: Added compute recommendation at the beginning of each notebook

File: genai-anomaly-detection/_resources/0_setup.py
Comment (kwulffert23) [inline genai-anomaly-detection/_resources/0_setup.py:1]: @marzervou added ingestion logic to work with Unity Catalog throughout the pca and the ml+genai notebooks

File: genai-anomaly-detection/1_anomaly_detection_pca.py (hunk 3 above)
Comment (kwulffert23) [reply genai-anomaly-detection/1_anomaly_detection_pca.py]: Good point, I added a logic to load the raw data into a volume in a catalog and then we use that catalog throughout the other notebooks

**Review by QuentinAmbard — APPROVED** [review]

**Comment by QuentinAmbard** [comment]
@kwulffert23 could you add a README in main folder explainin what this is about ? A small description with a link to the blog post when it's there?
Thanks

**Pull request merged**